### PR TITLE
Non-zero exit status in case of failure

### LINKED ISSUE #502: Return correct error code when error occurred like GNU ls
from https://github.com/Peltoche/lsd/issues/498 I found that `lsd` will not return an error code besides 0 even error occurred

we should implement this like the `GNU ls`

## Expected behavior
GNU ls output:

```
λ \ls a
ls: a: No such file or directory

✗ echo $?
1
```

## Actual behavior
lsd:

```
λ ls a
lsd: a: No such file or directory (os error 2).

λ echo $?
0
```

### PR DESCRIPTION
**Pull request by zarkone**
Mimics the value of exit status from GNU ls in case of failure:

> Exit status:
> 0      if OK,
> 1      if minor problems (e.g., cannot access subdirectory),
> 2      if serious trouble (e.g., cannot access command-line argument).

Closes #502

---
#### TODO

- [ ] Use `cargo fmt`
- [ ] Add necessary tests
- [ ] Add changelog entry

### CODE REVIEW

**Comment by zarkone** [comment]
Hello! 

I've noticed this "good first issue" label and decided give it a try. I haven't covered all the cases (including tests) -- metadata fails are not handled -- but decided to push it as is to ask if it is a good approach to return status like this. 

Thanks!

**Comment by zwpaper** [comment]
Hi @zarkone, thanks for the implementation! you have almost done the work.

maybe some updates like below could make further exit code cases easier to implement:

1. create a struct, `ExitStatus` for example, containing the code and a vector of errors
2. add a macro or fn like the `print_error!` to add error code and error to the `ExitStatus`
3. print all the errors in the main and exit with the code

what's your idea? @meain

**Comment by zarkone** [comment]
@zwpaper thanks for your feedback! Please check this recent commit https://github.com/Peltoche/lsd/pull/536/commits/3bca4047097b487b8b6675a5e7405ce56b1a71be if I understood your approach correct. If it is the case, I'll proceed with mod.rs and tests.

**Comment by zarkone** [comment]
BTW, I've noticed that `Meta::recurse_into` from `mod.rs` never returns an `Error`.. I think because of this [Err branch from recurse in fetch](https://github.com/zarkone/lsd/blob/502-file-not-found-exit-status/src/core.rs#L124)  never gets executed if you test it with `400` subdir and `--tree`.  

Am I missing something?

**Comment by meain** [comment]
@zwpaper  I don't think we should collect errors into a `vec` and print them in main. From what I can see, gnu ls just seems to print the errors as it goes through the directories. I think we can stick to that itself(especially useful in case of `tree` or other recursive items). Other than that, lets have the error returned as you mentioned in an `ExitStatus` enum, but that can be a simple enum just for clarity and convert that to the proper exit codes in `main`.

**Comment by zarkone** [comment]
@meain this is a good observation! And even more -- GNU `ls` doesn't collect/buffer anything anything at all -- it prints all the info right away. `lsd`, on the contrary, uses buffered approach.

 It actually brings quite a significant delay if you compare `ls` and `lsd` on a folder with big-enough amount of file / recursion depth: `ls` will start printing right away while `lsd` will have a noticeable silent delay before printing. 

With the current buffered approach of `lsd`, in case of any access errors, you'll see this errors first -- before printing any kind of fileinfo. And then, after delay of buffering, actual fileinfo will be printed. Erorrs and fileinfo output will be separated from eachother, which is not the case for GNU `ls`.

To summarize, I think `lsd` should either buffer everything or not buffer anything at all, to be consistent and not to produce confusing output (which I described in a paragraph above).

**Comment by meain** [comment]
I think in `lsd` we are printing the errors as we go while printing the actual output only at the end. But if we were to collect the errors and print it out at the end, the only change is that instead of printing the errors at the top, we will be print errors at the bottom which does not help much.

Also if we were to do something like:
```
ls existing-file1 non-existing-file existing-file2
```

gnu ls throws the error at top anyways.

```
ls: cannot access 'non-existing-file': No such file or directory
existing-file1  existing-file2
```

I don't personally see any use for collecting the errors into a vec as of now.

**Comment by zarkone** [comment]
@meain sounds good, thanks for the feedback! 

How about the other points? About error propagation from mod.rs and the implementation in general?

**Comment by zwpaper** [comment]
as I tested in `ubuntu 18.04` we can see `GNU ls` cached the error msg and show together like this:

test case, return 0:

```shell-session
lstest # ls -l
total 24
drwxr-xr-x 2 root root 4096 Jul 27 10:58 2
lrwxrwxrwx 1 root root    1 Jul 27 11:03 3 -> 3
drwxr-xr-x 2 root root 4096 Jul 27 10:58 4
drwxr-xr-x 2 root root 4096 Jul 27 10:58 6
drwxr-xr-x 2 root root 4096 Jul 27 10:58 a
lrwxrwxrwx 1 root root    9 Jul 27 11:05 b -> non-exist
drwxr-xr-x 2 root root 4096 Jul 27 10:58 c
drwxr-xr-x 2 root root 4096 Jul 27 10:58 e
```

symbol link dereference error, return 1:
```shell-session
lstest # ls -l -L
ls: cannot access '3': Too many levels of symbolic links
ls: cannot access 'b': No such file or directory
total 24
drwxr-xr-x 2 root root 4096 Jul 27 10:58 2
l????????? ? ?    ?       ?            ? 3
drwxr-xr-x 2 root root 4096 Jul 27 10:58 4
drwxr-xr-x 2 root root 4096 Jul 27 10:58 6
drwxr-xr-x 2 root root 4096 Jul 27 10:58 a
l????????? ? ?    ?       ?            ? b
drwxr-xr-x 2 root root 4096 Jul 27 10:58 c
drwxr-xr-x 2 root root 4096 Jul 27 10:58 e
```

symbol link dereference error, and non existed files, return 1:

```shell-session
lstest # ls -l -L . 5 d
ls: cannot access '5': No such file or directory
ls: cannot access 'd': No such file or directory
.:
ls: cannot access '3': Too many levels of symbolic links
ls: cannot access 'b': No such file or directory
total 24
drwxr-xr-x 2 root root 4096 Jul 27 10:58 2
l????????? ? ?    ?       ?            ? 3
drwxr-xr-x 2 root root 4096 Jul 27 10:58 4
drwxr-xr-x 2 root root 4096 Jul 27 10:58 6
drwxr-xr-x 2 root root 4096 Jul 27 10:58 a
l????????? ? ?    ?       ?            ? b
drwxr-xr-x 2 root root 4096 Jul 27 10:58 c
drwxr-xr-x 2 root root 4096 Jul 27 10:58 e
```

what's more, the `GNU ls` should also cache the result, it has to because it would rearrange the result for sorting and aligning, right?

**Comment by zwpaper** [comment]
I also noticed that `GNU ls` would print errors before each folder result:

```shell-session
lstest # ls -l -L -R .
.:
ls: cannot access '3': Too many levels of symbolic links
ls: cannot access 'b': No such file or directory
total 24
drwxr-xr-x 2 root root 4096 Jul 27 11:24 2
l????????? ? ?    ?       ?            ? 3
drwxr-xr-x 2 root root 4096 Jul 27 10:58 4
drwxr-xr-x 2 root root 4096 Jul 27 10:58 6
drwxr-xr-x 2 root root 4096 Jul 27 10:58 a
l????????? ? ?    ?       ?            ? b
drwxr-xr-x 2 root root 4096 Jul 27 10:58 c
drwxr-xr-x 2 root root 4096 Jul 27 10:58 e

./2:
ls: cannot access './2/4': Not a directory
total 0
-rw------- 1 root root 0 Jul 27 10:58 2
l????????? ? ?    ?    ?            ? 4
```

**Comment by meain** [comment]
@zwpaper Yeah, that does make sense. Since we will not be sorting across multiple directories, I don't think we have buffer all the output, but rather just per folder. But let us not get that done in this PR.  We could keep that as a separate PR.

@zarkone Yeah, you are right. As of now we are not returning any errors. Mostly it is just print the error and continue, but we can change that to return the `ExitCode` enum.

---

One other thing that I noticed when trying out what @zarkone mentioned is that `tree` does not recurse into the directory if the perm is `400`, but `lsd` as of now does. I believe we are currently emulating the behaviour of `ls -R` for tree which actually does recurse into the dirs.

Setup:
```
$ mkdir -p a/{b,c}
$ touch a/b/one a/c/two
$ chmod 400 a/b
```

`tree` (no errors here)
```
$ tree
.
└── a
    ├── b
    └── c
        └── two

3 directories, 1 file
```

`lsd --tree`
```
$ ll --tree
lsd: ./a/b/one: Permission denied (os error 13).

drwxr-xr-x  96B 42 seconds ago .
drwxr-xr-x 128B 42 seconds ago └── a
dr--------  96B now                ├── b
drwxr-xr-x  96B 35 seconds ago     └── c
.rw-r--r--   0B now                    └── two
```

`lsd -R`
```
$ ll -R a
lsd: a/b/one: Permission denied (os error 13).

dr-------- 96B 10 minutes ago b
drwxr-xr-x 96B 11 minutes ago c

a/b:

a/c:
.rw-r--r-- 0B 10 minutes ago two
```

**Comment by zwpaper** [comment]
ok, it make sense to me now printing the error message without caching it.

as the `tree` and `400` part, I would stand for the current for the option printing error.

1. `ls -R` would print the same error
2. the error message would help to verify there are files but not accessible

what's more, it seems that `ls -R` would print the filename with question mark in the unaccessible part:

```shell-session
$ ls -l -R
.:
total 4
drwxrwxr-x 4 ubuntu ubuntu 4096 Jul 28 10:41 a

./a:
total 8
dr-------- 2 ubuntu ubuntu 4096 Jul 28 10:41 b
drwxrwxr-x 2 ubuntu ubuntu 4096 Jul 28 10:41 c

./a/b:
ls: cannot access './a/b/one': Permission denied
total 0
-????????? ? ? ? ?            ? one

./a/c:
total 0
-rw-rw-r-- 1 ubuntu ubuntu 0 Jul 28 10:41 two
```

---

we should create 2 other issues for those 2 discuss

**Comment by zarkone** [comment]
> @zarkone Yeah, you are right. As of now we are not returning any errors. Mostly it is just print the error and continue, but we can change that to return the ExitCode enum.

@meain thanks, sounds good! 

@zwpaper, @meain:
Thanks for your investigations -- I was under impression that `ls` prints everything without any caching but It seems like I didn't red `ls` sources with enough attention. 

I'll remove error messages buffering, leaving only status code as an internal state of `core` then.

I also agree with @zwpaper that those issues are related, but out of scope of this PR -- just wanted to share my findings. I'm happy to jump on it as well after finishing this one -- if you feel it makes sense.

**Comment by zarkone** [comment]
@zwpaper @meain updated the PR -- kindly ask you to check my yet unfinished code if it is a good direction overall.

I feel that some tests and some code at least in `mod.rs` is still missing -- but if what I have now looks good than I'll proceed further with it.

File: src/core.rs [hunk 1]
```diff
@@ -20,6 +21,7 @@ pub struct Core {
     //display: Display,
     colors: Colors,
     sorters: Vec<(SortOrder, sort::SortFn)>,
+    exit_code: ExitCode,
```
Comment (meain) [inline src/core.rs:24]: Don't think we should initialize it here. We can just have a return from the recursive parse.

File: src/core.rs [hunk 2]
```diff
@@ -68,18 +70,23 @@ impl Core {
...
-    pub fn run(self, paths: Vec<PathBuf>) {
+    pub fn run(&mut self, paths: Vec<PathBuf>) {
         let mut meta_list = self.fetch(paths);
 
         self.sort(&mut meta_list);
-        self.display(&meta_list)
+        self.display(&meta_list);
+    }
+
+    pub fn exit(&self) {
```
Comment (meain) [inline src/core.rs:85]: Don't think we need this. We can just call `std::process::exit` directly.

File: src/core.rs [hunk 3]
```diff
@@ -92,6 +99,7 @@ impl Core {
                 Ok(meta) => meta,
                 Err(err) => {
                     print_error!("{}: {}.", path.display(), err);
+                    self.exit_code.set_if_greater(ExitCode::MajorIssue);
```
Comment (meain) [inline src/core.rs:102]: Instead of having exit code be a field in `Core` and pass into fetch. It can just be a local value that is initialized here and then returned back to the run function similar to what you have done for `recurse_into`.

File: src/core.rs [hunk 4]
```diff
@@ -68,18 +70,23 @@ impl Core {
...
-        self.display(&meta_list)
+        self.display(&meta_list);
+    }
+
+    pub fn exit(&self) {
+        std::process::exit(self.exit_code as i32);
     }
 
-    fn fetch(&self, paths: Vec<PathBuf>) -> Vec<Meta> {
+    fn fetch(&mut self, paths: Vec<PathBuf>) -> Vec<Meta> {
```
Comment (meain) [inline src/core.rs:89]: These `mut` can be removed.

File: src/main.rs [hunk 5]
```diff
@@ -115,7 +128,9 @@ fn main() {
         Config::default()
     };
     let flags = Flags::configure_from(&matches, &config).unwrap_or_else(|err| err.exit());
-    let core = Core::new(flags);
+    let mut core = Core::new(flags);
 
     core.run(inputs);
+
+    core.exit();
```
Comment (meain) [inline src/main.rs:135]: `core.run` can return the `ExitStatus` and we can just directly call `std::process::exit` here at the end.

**Review by meain — CHANGES_REQUESTED** [review]
Hey, just a quick change on how you approach this. I don't think `ExitStatus` is something that we could add into `Core`. I think it would be better if `ExitStatus` stays separate and we just return it  from `fetch`/`run` at the end of mutating a value inside `self`.

**Comment by meain** [comment]
Closing in favor of https://github.com/Peltoche/lsd/pull/739 . Thanks for starting the work on this PR :D
